### PR TITLE
mod hue and clamp saturation and luminance

### PR DESCRIPTION
**Pull request by greggman**
This makes it similar to rotation. With rotation you can do things
like

```
someObj.rotation.x += speed * deltaTime;
```

You don't have to keep the result in the 0 to 2*PI or -PI to PI range.

Similarly by modding hue you can cycle through hues with

```
someObj.hue += speed * deltaTime;
someColor.setHSL(someObj.hue, 1, 0.5);
```

That helps several use cases like tweening libaries

```
tween.fromTo(someObj, 5, { hue: 0 }, { hue: 5 });
```

It's also useful for random color selection from a given hue, saturation, luminance

```
var hsl = someColor.getHSL();
hsl.h += Math.random();
hsl.s += Math.random() * 0.2 - 0.1;  // make it slightly more or less saturated
someColor.setHSL(hsl.h, hsl.s, hsl.l);
```

### CODE REVIEW

File: src/math/Color.js [hunk 1]
```diff
@@ -62,39 +62,54 @@ THREE.Color.prototype = {
 
 	},
 
-	setHSL: function ( h, s, l ) {
+	setHSL: function() {
+		// comupute euclidian module of x % n
+		function emod ( x, n ) {
+			return x >= 0 ? ( x % n ) : ( ( n - ( -x % n ) ) % n );
```
Comment (WestLangley) [inline src/math/Color.js]: I believe the common fix for this is:

    function mod2( n, m ) {

        return ( ( n % m ) + m ) % m;

    }

File: src/math/Color.js [hunk 2]
```diff
@@ -62,39 +62,54 @@ THREE.Color.prototype = {
...
+		}
 
-		// h,s,l ranges are in 0.0 - 1.0
+		function clamp01 ( v ) {
+			return Math.min( 1, Math.max( 0, v ));
```
Comment (WestLangley) [inline src/math/Color.js]: You can just use the existing method `THREE.Math.clamp( v, 0, 1 )`.

File: src/math/Color.js [hunk 3]
```diff
@@ -62,39 +62,54 @@ THREE.Color.prototype = {
...
+		return function ( h, s, l ) {
 
-			this.r = this.g = this.b = l;
+			// h,s,l ranges are in 0.0 - 1.0
+			h = emod( h, 1 );
+			s = clamp01( s );
+			l = clamp01( l );
 
-		} else {
+			if ( s <= 0 ) {
```
Comment (WestLangley) [inline src/math/Color.js]: Probably should be `if ( s === 0 )`, since it can't be negative after it has been clamped.

**Comment by WestLangley** [comment]
You mean lightness not luminance.

I am not sure what to think about this... None of the parameters should be negative. I am not sure we should accommodate such values, but what you are proposing is not unreasonable, either.

**Comment by greggman** [comment]
Do you think `Math.sin` should be limited to +/- Math.PI?

**Comment by mrdoob** [comment]
I like the change :+1:

**Comment by WestLangley** [comment]
> Do you think Math.sin should be limited to +/- Math.PI?

Of course not. Radians is defined on the real line.

According to https://en.wikipedia.org/wiki/HSL_and_HSV, hue has units of degrees, and hence, allowing `hue` to wrap, as you have proposed, makes sense.

So, do you want to modify `THREE.ColorConverter.setHSV()`, too?

**Comment by mrdoob** [comment]
Many thanks guys!

**Pull request merged**